pkg/storage/cacher.go

```diff
@@ -859,10 +859,19 @@ func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
 	case !curObjPasses && oldObjPasses:
 		watchEvent = watch.Event{Type: watch.Deleted, Object: object}
 	}
+
+	// We don't want to block on c.result if c.done is already closed.
+	//
+	// TODO: This doesn't protect with the situation when c.result is
+	// already full and we are handing on it and in the meantime c.done
+	// is being closed and noone will clear the result channel.
+	// However, we can't heve select of two cases, as this may lead to
+	// delivering random part of events after c.done is closed but we
+	// are still sending them for a moment.
 	select {
-	case c.result <- watchEvent:
-	// don't block on c.result if c.done is closed
 	case <-c.done:
+	default:
```

Comment:
@wojtek-t 

So I read https://github.com/kubernetes/kubernetes/issues/39072. Now I understand the problem. The comment is very misleading. We need to fix it.

Also how about

``` go
c.Lock
stopped := c.stopped
c.Unlock
if stopped {
    return
}

// keep previous logic
```

not sure if the lock adds up latency. but it would help the race between close(done) and the default case here. or we might end up with leaking routine again.